### PR TITLE
replace all trit representations from int[] to byte[]

### LINKED ISSUE #799: Make the hash class take up less memory
### Description
Make the hash class instance take less memory. 

### Motivation
We want to reduce the memory footprint of the class by a factor of 4. 

### Requirements
- [ ] trits array type is of byte[] and resolves errors that rise

### PR DESCRIPTION
**Pull request by alon-e**
# Description
Replace all trit representations from int[] to byte[]
Currently, Trits (-1,0,1) are stored in integer arrays, which waste memory (and compute to GC).
This PR replaces all usage trit representations from `int[]` to `byte[]`.

Fixes #799 

## Type of change

- Enhancement (a non-breaking change which adds functionality)

# How Has This Been Tested?
- Fixed unit tests
- Syncing a node with brach - improved memory and CPU footprint be > 50%

# Checklist:

- [x] My code follows the style guidelines for this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/test/java/com/iota/iri/controllers/TransactionViewModelTest.java [hunk 1]
```diff
@@ -106,34 +102,32 @@ public void update() throws Exception {
...
             System.arraycopy(blanks, 0, trits, TransactionViewModel.TRUNK_TRANSACTION_TRINARY_OFFSET-blanks.length, blanks.length);
             System.arraycopy(blanks, 0, trits, TransactionViewModel.BRANCH_TRANSACTION_TRINARY_OFFSET-blanks.length, blanks.length);
             System.arraycopy(blanks, 0, trits, TransactionViewModel.BRANCH_TRANSACTION_TRINARY_OFFSET + TransactionViewModel.BRANCH_TRANSACTION_TRINARY_SIZE-blanks.length, blanks.length);
-            TransactionViewModel transactionViewModel = new TransactionViewModel(trits);
-            transactionViewModel.store();
-            assertArrayEquals(transactionViewModel.trits(), TransactionViewModel.fromHash(transactionViewModel.getHash()).trits());
+            Hash hash = getRandomTransactionHash();
+            TransactionViewModel transactionViewModel = new TransactionViewModel(trits, hash);
+            transactionViewModel.store(tangle);
+            assertArrayEquals(transactionViewModel.trits(), TransactionViewModel.fromHash(tangle, transactionViewModel.getHash()).trits());
```
Comment (alon-e) [inline src/test/java/com/iota/iri/controllers/TransactionViewModelTest.java]: re-enabled test and fixed, could move to another PR if you want.

File: src/test/java/com/iota/iri/controllers/TransactionViewModelTest.java [hunk 2]
```diff
@@ -106,34 +102,32 @@ public void update() throws Exception {
...
+            byte[] trits = getRandomTransactionTrits();
+            System.arraycopy(new byte[TransactionViewModel.VALUE_TRINARY_SIZE], 0, trits, TransactionViewModel.VALUE_TRINARY_OFFSET, TransactionViewModel.VALUE_TRINARY_SIZE);
             Converter.copyTrits(seed.nextLong(), trits, TransactionViewModel.VALUE_TRINARY_OFFSET, TransactionViewModel.VALUE_USABLE_TRINARY_SIZE);
-            TransactionViewModel transactionViewModel = new TransactionViewModel(trits);
-            transactionViewModel.store();
-            assertArrayEquals(transactionViewModel.getBytes(), TransactionViewModel.fromHash(transactionViewModel.getHash()).getBytes());
+            Hash hash = getRandomTransactionHash();
+            TransactionViewModel transactionViewModel = new TransactionViewModel(trits, hash);
+            transactionViewModel.store(tangle);
+            assertArrayEquals(transactionViewModel.getBytes(), TransactionViewModel.fromHash(tangle, transactionViewModel.getHash()).getBytes());
```
Comment (alon-e) [inline src/test/java/com/iota/iri/controllers/TransactionViewModelTest.java]: re-enabled test and fixed, could move to another PR if you want.

File: src/test/java/com/iota/iri/controllers/TransactionViewModelTest.java [hunk 3]
```diff
@@ -396,47 +390,40 @@ public void testManyTXInDB() throws Exception {
     private Transaction getRandomTransaction(Random seed) {
         Transaction transaction = new Transaction();
 
-        int[] trits = Arrays.stream(new int[TransactionViewModel.SIGNATURE_MESSAGE_FRAGMENT_TRINARY_SIZE]).map(i -> seed.nextInt(3)-1).toArray();
+        byte[] trits = new byte[TransactionViewModel.SIGNATURE_MESSAGE_FRAGMENT_TRINARY_SIZE];
```
Comment (alon-e) [inline src/test/java/com/iota/iri/controllers/TransactionViewModelTest.java]: `getRandomTransaction` impl. appears a few times, better to move to a Utils class.
but not part of this scope.

File: src/test/java/com/iota/iri/controllers/TransactionViewModelTest.java [hunk 4]
```diff
@@ -396,47 +390,40 @@ public void testManyTXInDB() throws Exception {
...
-        int[] trits = getRandomTransactionTrits();
+    public static byte[] getRandomTransactionWithTrunkAndBranch(Hash trunk, Hash branch) {
+        byte[] trits = getRandomTransactionTrits();
         System.arraycopy(trunk.trits(), 0, trits, TransactionViewModel.TRUNK_TRANSACTION_TRINARY_OFFSET,
                 TransactionViewModel.TRUNK_TRANSACTION_TRINARY_SIZE);
         System.arraycopy(branch.trits(), 0, trits, TransactionViewModel.BRANCH_TRANSACTION_TRINARY_OFFSET,
                 TransactionViewModel.BRANCH_TRANSACTION_TRINARY_SIZE);
         return trits;
     }
-    public static int[] getRandomTransactionWithTrunkAndBranchValidBundle(Hash trunk, Hash branch) {
```
Comment (alon-e) [inline src/test/java/com/iota/iri/controllers/TransactionViewModelTest.java]: removed b/c not used.

File: src/test/java/com/iota/iri/controllers/TransactionViewModelTest.java (hunk 3 above)
Comment (GalRogozinski) [reply src/test/java/com/iota/iri/controllers/TransactionViewModelTest.java]: 👍

File: src/test/java/com/iota/iri/controllers/TransactionViewModelTest.java [hunk 5]
```diff
@@ -396,47 +390,40 @@ public void testManyTXInDB() throws Exception {
...
-        return trits;
-    }
-    public static int[] getRandomTransactionTrits() {
-        return Arrays.stream(new int[TransactionViewModel.TRINARY_SIZE]).map(i -> seed.nextInt(3)-1).toArray();
+        for(int i = 0; i < out.length; i++) {
+            out[i] = (byte) (seed.nextInt(3) - 1);
+        }
+
+        return out;
     }
```
Comment (GalRogozinski) [inline src/test/java/com/iota/iri/controllers/TransactionViewModelTest.java]: I agree that this is more reasdable than the one liner.
Another option was to not change the code and just add new lines before each aggregate call.

Anyhows, both variations are fine with me.

File: src/main/java/com/iota/iri/controllers/TransactionViewModel.java [hunk 6]
```diff
@@ -85,29 +85,29 @@ public TransactionViewModel(final Transaction transaction, final Hash hash) {
...
+            transaction.arrivalTime = 0;
+        }
+        else {
+            transaction = new Transaction();
+            transaction.bytes = new byte[SIZE];
+            System.arraycopy(trits, 0, transaction.bytes, 0, SIZE);
+            this.hash = hash;
+            weightMagnitude = this.hash.trailingZeros();
+            transaction.type = FILLED_SLOT;
+        }
```
[the remark this answers is not in the document]
Comment (alon-e) [reply src/main/java/com/iota/iri/controllers/TransactionViewModel.java]: I'll open an issue on this. but it shouldn't block the goodness from this PR imo.

File: src/main/java/com/iota/iri/model/Hash.java [hunk 7]
```diff
@@ -34,59 +34,57 @@ private ByteSafe(byte[] bytes) {
...
+        if(sourceSize < SIZE_IN_TRITS) {
+            byte[] dest = new byte[SIZE_IN_BYTES];
+            System.arraycopy(source, sourceOffset, dest, 0, sourceSize - sourceOffset > source.length ? source.length - sourceOffset : sourceSize);
+            this.byteSafe = new ByteSafe(dest);
+        } else {
+            byte[] dest = new byte[SIZE_IN_TRITS];
+            System.arraycopy(source, sourceOffset, dest, 0, dest.length);
+            this.tritSafe = new TritSafe(dest);
+        }
     }
```
[the remark this answers is not in the document]
Comment (alon-e) [reply src/main/java/com/iota/iri/model/Hash.java]: I'll open an issue on this. but it shouldn't block the goodness from this PR imo.
